### PR TITLE
fix(multi-tenancy): update /state on delete namespace

### PR DESCRIPTION
**Pull request by NamanJain8**
Delete namespace is sent to zero after alpha groups delete the namespace. Zero proposes the changes internally then.
Membership state is updated to reflect the deleted namespace, i.e., all the predicates of the deleted 
namespace are removed from the `/state`.

---
This change is [Reviewable](https://reviewable.io/reviews/dgraph-io/dgraph/7463)

### CODE REVIEW

**Review by ahsanbarkati — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/dgraph-io/dgraph/7463)* status: 0 of 8 files reviewed, 1 unresolved discussion (waiting on @manishrjain, @martinmr, @NamanJain8, and @vvbalaji-dgraph)

___
*[dgraph/cmd/zero/zero.go, line 610 at r1](https://reviewable.io/reviews/dgraph-io/dgraph/7463#-MU7jJix6jm5oMQ9TVtg:-MU7jJiyDxnWXa53nTrK:b-6bdm1j) ([raw file](https://github.com/dgraph-io/dgraph/blob/7761e5516abcbb225d70b861d341b6e9a4bfb4cf/dgraph/cmd/zero/zero.go#L610)):*
> ```Go
> 	g := new(errgroup.Group)
> 	for gid := range state.Groups {
> 		if gid == 0 {
> ```

You probably should create a new copy of gid within the loop.
https://golang.org/doc/faq#closures_and_goroutines

**Review by NamanJain8 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/dgraph-io/dgraph/7463)* status: 0 of 8 files reviewed, 1 unresolved discussion (waiting on @ahsanbarkati, @manishrjain, @martinmr, and @vvbalaji-dgraph)

___
*[dgraph/cmd/zero/zero.go, line 610 at r1](https://reviewable.io/reviews/dgraph-io/dgraph/7463#-MU7jJix6jm5oMQ9TVtg:-MU8XgIP411p62khEHRo:b-i21bbo) ([raw file](https://github.com/dgraph-io/dgraph/blob/7761e5516abcbb225d70b861d341b6e9a4bfb4cf/dgraph/cmd/zero/zero.go#L610)):*
<details><summary><i>Previously, ahsanbarkati (Ahsan Barkati) wrote…</i></summary><blockquote>

You probably should create a new copy of gid within the loop.
https://golang.org/doc/faq#closures_and_goroutines
</blockquote></details>

Thanks. Done.

File: worker/multi_tenancy_ee.go [hunk 1]
```diff
@@ -39,43 +34,12 @@ func (w *grpcWorker) DeleteNamespace(ctx context.Context,
...
 func ProcessDeleteNsRequest(ctx context.Context, ns uint64) error {
-	// Update the membership state to get the latest mapping of groups to predicates.
-	if err := UpdateMembershipState(ctx); err != nil {
-		return errors.Wrapf(err, "Failed to update membership state while deleting namesapce")
-	}
-
-	state := GetMembershipState()
-	g := new(errgroup.Group)
-
-	for gid := range state.Groups {
```
Comment (jarifibrahim) [inline worker/multi_tenancy_ee.go:50]: Let's not route the delete request through zero.
We can process the delete namespace request in alpha and then return it to zero.

Comment (NamanJain8) [reply worker/multi_tenancy_ee.go]: Done.

**Review by jarifibrahim — CHANGES_REQUESTED** [review]
> Dismissed @ahsanbarkati from a discussion.
> *[Reviewable](https://reviewable.io/reviews/dgraph-io/dgraph/7463)* status: 0 of 9 files reviewed, 2 unresolved discussions (waiting on @manishrjain, @martinmr, @NamanJain8, and @vvbalaji-dgraph)

___
*[dgraph/cmd/zero/zero.go, line 597 at r4](https://reviewable.io/reviews/dgraph-io/dgraph/7463#-MUEmMpa-k1xrmQHIiHY:-MUEmMpb9qfT8rrxUlHj:blsg16o) ([raw file](https://github.com/dgraph-io/dgraph/blob/3506d824080506b600ba829eeb9b23861edacb71/dgraph/cmd/zero/zero.go#L597)):*
> ```Go
> func (s *Server) DeleteNamespace(ctx context.Context, in *pb.DeleteNsRequest) (*pb.Status, error) {
> 	err := s.Node.proposeAndWait(ctx, &pb.ZeroProposal{DeleteNs: in})
> 	return &pb.Status{}, err
> ```

@namanjain8  Why return an empty status? Don't you want to return a success status if the error is not nil?
___
*[worker/multi_tenancy_ee.go, line 64 at r4](https://reviewable.io/reviews/dgraph-io/dgraph/7463#-MUEmxXu7sRzbYUwKkKa:-MUEmxXvCbiDPEQh-jLs:b2sstgy) ([raw file](https://github.com/dgraph-io/dgraph/blob/3506d824080506b600ba829eeb9b23861edacb71/worker/multi_tenancy_ee.go#L64)):*
> ```Go
> 
> 	// Now propose the change to zero.
> 	return x.RetryUntilSuccess(100, 10*time.Second, func() error {
> ```

You'll retry 100 times and each request will be sent after 10 seconds? That would be very very slow.

You could do more like 10 requests and each one after every 100 millisecond.

**Review by jarifibrahim — APPROVED** [review]
:lgtm: Please address the comments before merging.

> *[Reviewable](https://reviewable.io/reviews/dgraph-io/dgraph/7463#-:-MUEo83x0_5gg54bX5oT:buhblf4)* status: 0 of 9 files reviewed, 2 unresolved discussions (waiting on @manishrjain, @martinmr, @NamanJain8, and @vvbalaji-dgraph)

**Review by NamanJain8 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/dgraph-io/dgraph/7463)* status: 0 of 9 files reviewed, 2 unresolved discussions (waiting on @jarifibrahim, @manishrjain, @martinmr, and @vvbalaji-dgraph)

___
*[dgraph/cmd/zero/zero.go, line 597 at r4](https://reviewable.io/reviews/dgraph-io/dgraph/7463#-MUEmMpa-k1xrmQHIiHY:-MUEr1nJBDBx0Ji6nRQd:b-b99yqy) ([raw file](https://github.com/dgraph-io/dgraph/blob/3506d824080506b600ba829eeb9b23861edacb71/dgraph/cmd/zero/zero.go#L597)):*
<details><summary><i>Previously, jarifibrahim (Ibrahim Jarif) wrote…</i></summary><blockquote>

@namanjain8  Why return an empty status? Don't you want to return a success status if the error is not nil?
</blockquote></details>

We are not utilizing the status. We need only the error. GRPC service must return a Typed message.

We can use `Empty` type https://github.com/gogo/protobuf/blob/deb6fe8ca7c6d06584bfbd40ca407bf69d9fd2aa/protobuf/google/protobuf/empty.proto#L43-L52 

I used the approach as in Restore to keep consistency.
___
*[worker/multi_tenancy_ee.go, line 64 at r4](https://reviewable.io/reviews/dgraph-io/dgraph/7463#-MUEmxXu7sRzbYUwKkKa:-MUEt4-G3Pw90AEdbAib:b-i21bbo) ([raw file](https://github.com/dgraph-io/dgraph/blob/3506d824080506b600ba829eeb9b23861edacb71/worker/multi_tenancy_ee.go#L64)):*
<details><summary><i>Previously, jarifibrahim (Ibrahim Jarif) wrote…</i></summary><blockquote>

You'll retry 100 times and each request will be sent after 10 seconds? That would be very very slow.

You could do more like 10 requests and each one after every 100 millisecond.
</blockquote></details>

Thanks. Done.

**Pull request merged**